Adapt to changes needed for `gardener-node-agent`

rfranzke opened the pull request:
**How to categorize this PR?**
/area os
/kind enhancement

**What this PR does / why we need it**:
This PR adapts this extension to changes needed for `gardener-node-agent`, see https://github.com/gardener/gardener/pull/8647 for more information.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/8023

**Special notes for your reviewer**:
Kindly use the dedicated commits for an easier review process.

/cc @oliver-goetz @majst01 @Gerrit91 

**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|noteworthy|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```noteworthy operator
This extension is now prepared to run with an enabled `UseGardenerNodeAgent` feature gate.
```
```noteworthy operator
It is assumed that `gardenlet`'s `UseGardenerNodeAgent` is turned on by default if the feature gate is not explicitly set. Hence, make sure to use at least Gardener `v1.82` when using this extension version.
```

rfranzke committed:
Revendor g/g to v1.89.0

rfranzke committed:
Rework directory restructure

In the future, the `pkg/controller/operatingsystemconfig/generator` package can be deleted entirely

rfranzke committed:
Rework main entrypoint

- no longer depend on `oscommon` package

rfranzke committed:
Introduce OSC controller

- no longer use `oscommon` package
- prepare for new OSC contract when `UseGardenerNodeAgent` feature gate is enabled

majst01 commented:
Hi @rfranzke 

This is awsome, thanks so much, will take a while til we can test this.

rfranzke commented:
Why does the build fail? Running `golangci-lint run  -p bugs --timeout=3m ./...` locally on this branch doesn't yield any findings 👀

majst01 commented:
> Why does the build fail? Running `golangci-lint run -p bugs --timeout=3m ./...` locally on this branch doesn't yield any findings 👀

It says NewActuator is undefined ?

rfranzke commented:
I see this 😄 But all occurrences of `NewActuator` are fine, even `go run` and `go test` work, and local `golangci-lint` run does not find anything. I'm confused...

majst01 commented:
> I see this 😄 But all occurrences of `NewActuator` are fine, even `go run` and `go test` work, and local `golangci-lint` run does not find anything. I'm confused...

I check

majst01 commented:
I updated master with this PR: https://github.com/metal-stack/os-metal-extension/pull/34 which compiled nicely, please pull master and it should compile your PR again.

Guess its because the newer linter does not like go-1.20 ??

rfranzke committed:
Set `UseGardenerNodeAgent` option based on gardenlet feature gate

It is `true` by default (even when `UseGardenerNodeAgent` is not provided via `.Values.gardener.gardenlet.featureGates` in the Helm chart values).

rfranzke commented:
/hold
Let's wait with the merge for another revendoring with `gardener/gardener@v1.83.0` containing https://github.com/gardener/gardener/pull/8707

rfranzke commented:
/unhold
`gardener/gardener@v1.84` has been released today, and this PR has been adapted accordingly.

/cc @Gerrit91 @majst01 @oliver-goetz 
I think it's ready to be merged now. Please proceed with the review, thank you!

majst01 commented:
> /unhold `gardener/gardener@v1.84` has been released today, and this PR has been adapted accordingly.
> 
> /cc @Gerrit91 @majst01 @oliver-goetz I think it's ready to be merged now. Please proceed with the review, thank you!

Will do, thanks a lot

oliver-goetz commented on pkg/controller/operatingsystemconfig/actuator.go:
```diff
@@ -0,0 +1,151 @@
+	writeUnitsToDiskScript := operatingsystemconfig.UnitsToDiskScript(osc.Spec.Units)
+
+	script := `#!/bin/bash
+#Fix mis-configuration of dockerd
```
```suggestion
# Fix mis-configuration of dockerd
```
nit

oliver-goetz commented on pkg/controller/operatingsystemconfig/actuator.go:
```diff
@@ -0,0 +1,151 @@
+`
+
+	for _, unit := range osc.Spec.Units {
+		script += fmt.Sprintf(`systemctl enable '%s' && systemctl restart '%s'
```
```suggestion
		script += fmt.Sprintf(`systemctl enable '%s' && systemctl restart --no-block '%s'
```
You could add a `--no-block` as we did for other OS extensions ([ref](https://github.com/gardener/gardener-extension-os-gardenlinux/blob/b1d661d31368826551bf30da3df28115cbe2247b/pkg/controller/operatingsystemconfig/actuator.go#L122)). This would prevent issues that the creation of `gardener` user we introduce with https://github.com/gardener/gardener/pull/9077 could be blocked by an error in `gardener-node-init` systemd unit.

oliver-goetz reviewed (commented):
nice PR 😄

rfranzke committed:
Address PR review feedback

Gerrit91 commented:
/ok-to-build

Gerrit91 commented:
The controller cannot be deployed because the installation helm chart is now broken. 😅

```
    message: 'chart rendering process failed: render error in "os-metal/templates/deployment.yaml":
      template: os-metal/templates/deployment.yaml:34:52: executing "os-metal/templates/deployment.yaml"
      at <.Values.leaderElection.resourceLock>: nil pointer evaluating interface {}.resourceLock'
    reason: ChartCannotBeRendered
```

rfranzke commented:
> The controller cannot be deployed because the installation helm chart is now broken. 😅

Sorry for the delay, this should be fixed now. I've also rebased to `master` and incorporated the changes from #38 into this PR. Please check again :)

rfranzke commented:
Please note, there is one more addition:

> It is assumed that `gardenlet`'s `UseGardenerNodeAgent` is turned on by default if the feature gate is not explicitly set. Hence, make sure to use at least Gardener `v1.82` when using this extension version.

rfranzke commented:
Later, when this PR is merged, and then even later, after you upgraded to `gardener/gardener@v1.92` or higher, you can take a look at https://github.com/gardener/gardener-extension-os-gardenlinux/pull/161 for what needs to be done to remove the migration/legacy code.
